### PR TITLE
Update dumb_downstairs test to be smarter

### PR DESCRIPTION
**Pull request by leftwo**
Updated parts of the dummy_downstairs to allow for ping packets as well as different results from channels that might be either closed or empty.

### CODE REVIEW

File: .github/buildomat/jobs/build.sh [hunk 1]
```diff
@@ -20,6 +20,8 @@ rustc --version
 banner build
 ptime -m cargo build --verbose
 
+ptime -m cargo test test_successful_live_repair --verbose
```
Comment (jmpesp) [inline .github/buildomat/jobs/build.sh:23]: *remove before flight*

File: upstairs/src/dummy_downstairs_tests.rs [hunk 2]
```diff
@@ -156,102 +156,123 @@ pub(crate) mod protocol_test {
...
-                    version,
-                    upstairs_id: _,
-                    session_id: _,
-                    gen: _,
-                    read_only: _,
-                    encrypted: _,
-                    alternate_versions: _,
-                } => {
-                    info!(self.inner.log, "negotiate packet {:?}", packet);
+            loop {
```
Comment (jmpesp) [inline upstairs/src/dummy_downstairs_tests.rs:159]: currently, a ping can't happen before `HereIAm`, it only happens in the select loop after that packet has been sent.

File: upstairs/src/dummy_downstairs_tests.rs [hunk 3]
```diff
@@ -156,102 +156,123 @@ pub(crate) mod protocol_test {
...
-                .unwrap()
-                .unwrap();
-            match &packet {
-                Message::PromoteToActive {
-                    upstairs_id,
-                    session_id,
-                    gen,
-                } => {
-                    assert!(*gen == 1);
+            loop {
```
Comment (jmpesp) [inline upstairs/src/dummy_downstairs_tests.rs:201]: There should be a comment explaining the ping behaviour up here maybe, instead of in the match?

**Review by jmpesp — APPROVED** [review]
Sorry about the test fails!

File: upstairs/src/dummy_downstairs_tests.rs (hunk 2 above)
Comment (leftwo) [reply upstairs/src/dummy_downstairs_tests.rs]: Done

File: upstairs/src/dummy_downstairs_tests.rs (hunk 3 above)
Comment (leftwo) [reply upstairs/src/dummy_downstairs_tests.rs:195]: Done

**Comment by leftwo** [comment]
This makes it better (found some issues) but there is still some out there that prevents it from working
every time.  I'm going to merge this change but we should continue to iterate on whatever is ailing this test.

**Pull request merged**